Move `hash_bytes_into_chunks` from execution-engine to `Digest` type.

goral09 opened the pull request:
We want to reuse the chunking logic in more places, to chunk other data types that just trie, and they all have to follow the same scheme:
* be chunked before storing
* have chunks' IDs being computed (in the fetcher) in a way that is aligned with how the data is stored

By moving this method from execution engine to `Digest` type we're making it easier for developers to use the correct logic instead of writing a new version of it.

Closes https://github.com/casper-network/casper-node/issues/3265 and https://github.com/casper-network/casper-node/issues/3264

Commit:
Move hash_bytes_into_chunks to Digest type.

Commit:
Use merkelized hashing for execution results.

Also, instead of just hashing 32-byte hashes of the results we will hash
whole serialized representations. For more context, see comment on the
method.

Fraser999 commented on node/src/components/contract_runtime/operations.rs:
```diff
@@ -402,16 +402,23 @@ where
     results: &mut impl Iterator<Item = &'a ExecutionResult>,
 ) -> Result<Digest, BlockCreationError> {
-    let mut execution_results = vec![];
+    let mut execution_results_serialized = vec![];
```
I think we need to do something more like changing `results` to a `Vec<ExecutionResult>` and just call `let execution_results_bytes = results.to_bytes()...` - that way we get the length of the vector written to the output bytes first, which is what I think we'll need when serializing the actual execution results to send them.

goral09 replied:
We should get it here as well since we're calling `ToBytes` for each `ExecutionResult`. The input to `hash_bytes_into_*` is `execution_results_bytes` which is result of calling `ToBytes` on `Vec<Vec<u8>>` . Each individual element of the outer vector should be deserializable to `ExecutionResult` type (since that's what we serialize) and the outer vector should get its length serialized first as well.

I can change this to:
```rust
let execution_results_bytes = results.clone().to_bytes()
Digest::hash_bytes_into_chunks_if_necessary(&execution_results_bytes )
```
since they should be equivalent.

goral09 replied:
Fixed in [`51e7d25` (#3263)](https://github.com/casper-network/casper-node/pull/3263/commits/51e7d2555cbe6230bf85db1b3ce073378eac6cc5).

Commit:
Fix calculation of execution results' root hash.

The previous logic would "double serialize" the inner elements:
1. ExecutionResult -> Vec<u8>
2. When serializing outer vector we would prepend length of each inner
vectors before writing its bytes.

goral09 commented:
bors merge
